### DISCUSSION TITLE
Non-docker installation

### DISCUSSION THREAD

**Discussion by myOmikron** [kanidm/kanidm#1436] [General]
Hey guys,
I'd like to setup kanidm in a non-docker environment, the book starts with this section:

> NOTE Our preferred deployment method is in containers, and this documentation assumes you're running in docker. Kanidm will alternately run as a daemon/service, and server builds are available for multiple platforms if you prefer this option. You will

I'd be wiling to write documentation for a non-docker setup if you are interested in this option.

**Comment by Firstyear** [comment]
There are multiple reasons why we haven't documented for platforms.

* Distros tend to lag behind our upstream releases. Not always, but sometimes. Some distros are "more up to date" than others, but also there has already been a case where our support for a distro has had to be removed because of issues. So by prioritising containers we know it will "work everywhere for everyone". This will be even more important as we will likely start to prevent / remove the ability to upgrade from more than 1 or 2 releases prior. 
* If a distro ships Kanidm, and it's not a "rolling" distro, then we start to get wedged with supporting something in a package for far longer than we wanted (see debian as an example where software can become very old). We don't want to "screw over" our users for upgrades, we want a project/package lifecycle that is able to be detached from the running host.
* Distros tend to have subtle differences. These make it hard to document for and account for all the differences. We'd end up with a lot of "per-distro" docs which can make it harder for people to find what they need. 
* We do try to test the instructions in our manual. If we add more distros, that's more testing for us.

So our approach has been "let's do one thing, and one thing well". This is why I'd be hesitant to add non-docker setup instructions.

**Comment by yaleman** [comment]
It does go on to mention the following:

> You may need to adjust your example commands throughout this document to suit your desired server type if you choose not to use docker.

We could probably just mention that there's example systemd service files in `/examples/systemd` which *should* be enough for most folks?

**Reply by Firstyear** [reply]
The problem is that with the systemd setup vs docker you need to know to chown/mod stuff for the initial run of the db and the account recovery process is completely different. You can't "inject" commands into the systemd unit file to allow the initial recovery to occur for example.

And if you do the initial recovery first outside of systemd, all the files are owned by the wrong user so the service will fail to start or operation correctly. If we have selinux present there is also a possibility the files will be mislabeled too.

So it's not quite so simple to actually bypass this.

**Reply by yaleman** [reply]
Oh, I know... having dealt with it before it's a bit of a pain, but it's doable. The mention more would be "if you must, then here's a path... but we won't support it unless it's a bug in the code."

**Reply by Firstyear** [reply]
The other side of that is that by documenting/acknowledging it we're saying "it's an option" rather than "look, just use the container pls".

**Reply by hvisage** [reply]
Problem with the docker/containers, in my case I want a LXC container at most, or a simple single function VM, and docker just makes it much much more complex for me to manage than a simple VM/LXC

**Reply by hvisage** [reply]
oh, and dockers doesn't run in nicely in LXCs, and when you have a simple VM, to add docker becomes a nightmare with the extras I now have to download every time instead of a simple push (via Ansible) of the new binaries into the VM/LXC...

**Comment by myOmikron** [comment]
Hmm okay, yeah, I got your point. 

I'm on your side with packaging for distros, those are often outdated.

I've often seen documentation for building from source / download prebuilt packages for few specific distros, may that be an option?

And maybe a bit offtopic, have you considered providing a docker-compose file? 
At least from my perspective, it would simplify the current deployment configuration.

**Reply by Firstyear** [reply]
I'd want to see what the docker-compose setup looks like to create that deployment - in my experience docker-compose only works well for "stateless" applications and kanidm is the furthest away from this. But I'm open to that. 

PS: Another issue with distro packages is distros tend to compile without cpu optimisations, where our docker images have neonv8 or x86_64-v2 which are not insignificant improvements for us

**Reply by yaleman** [reply]
Here's my docker-compose.yml which I've been using to run Kanidm + RADIUS in prod for a while. 

It uses:

- the github-built images, but would work fine with any IIRC.
- host networking because I want to see the source IP and expose it on IPv6 and docker tends to be a jerk about both.



```docker-compose.yml
version: "3.5"

# This assumes a couple of things:
#
# 1. You're using letsencrypt certificates, and fullchain.pem and privkey.pem are in /opt/letsencrypt-certs/
# 2. You've got docker-compose installed (python -m pip install docker-compose)

services:
  kanidm:
    container_name: kanidmd
    restart: "always"
    image: ghcr.io/kanidm/kanidmd:devel
    volumes:
      - /opt/kanidm-container/data/:/var/lib/kanidm/
      - /etc/kanidm/:/etc/kanidm/
      - /etc/kanidm/kanidmd.toml:/data/server.toml
    network_mode: host
  radiusd:
    container_name: radiusd
    restart: "always"
    image: ghcr.io/kanidm/radius:devel
    volumes:
      - /opt/kanidm-container/radius.toml:/data/kanidm
      - /etc/kanidm/kanidmd.toml:/data/server.toml
      - /opt/kanidm-container/radius//:/certs/
    ports:
      - 1812:1812/udp
      - 1813:1813/udp
      - 1812:1812/tcp
      - 1813:1813/tcp
```
